base_layer/common_types/src/dammsum.rs

```diff
@@ -0,0 +1,210 @@
+// disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the
+// following disclaimer in the documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote
+// products derived from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+// INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+// USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use thiserror::Error;
+
+/// Calculates a checksum using the [DammSum](https://github.com/cypherstack/dammsum) algorithm.
+///
+/// This approach uses a dictionary whose size must be `2^k` for some `k > 0`.
+/// The algorithm accepts an array of arbitrary size, each of whose elements are integers in the range `[0, 2^k)`.
+/// The checksum is a single element also within this range.
+/// DammSum detects all single transpositions and substitutions.
+///
+/// Note that for this implementation, we add the additional restriction that `k == 8`.
+/// This is only because DammSum requires us to provide the coefficients for a certain type of polynomial, and
+/// because it's unlikely for the alphabet size to change for this use case.
+/// See the linked repository for more information, or if you need a different dictionary size.
+
+#[derive(Debug, Error, PartialEq)]
+pub enum ChecksumError {
+    #[error("Input data is too short")]
+    InputDataTooShort,
+    #[error("Invalid checksum")]
+    InvalidChecksum,
+}
+
+// Fixed for a dictionary size of `2^8 == 256`
+const COEFFICIENTS: [u8; 3] = [4, 3, 1];
+
+/// Compute the DammSum checksum for an array, each of whose elements are in the range `[0, 2^8)`
+pub fn compute_checksum(data: &Vec<u8>) -> u8 {
+    let mut mask = 1u8;
+
+    // Compute the bitmask (if possible)
+    for bit in COEFFICIENTS {
+        mask += 1u8 << bit;
+    }
+
+    // Perform the Damm algorithm
+    let mut result = 0u8;
+
+    for digit in data {
+        result ^= *digit; // add
+        let overflow = (result & (1 << 7)) != 0;
+        result <<= 1; // double
+        if overflow {
+            // reduce
+            result ^= mask;
+        }
+    }
+
+    result as u8
```

Comment:
Why do we need `result as u8` and not just `result` ? `result was of type `u8` all along.

Reply:
This is just an artifact of when `result` was a `u16` to hold the overflow. Will change.